### PR TITLE
Add Translation support to Module and ModuleStream

### PR DESCRIPTION
**Pull request by puiterwijk**
Signed-off-by: Patrick Uiterwijk <puiterwijk@redhat.com>

### CODE REVIEW

**Comment by puiterwijk** [comment]
Tests added

File: modulemd/v2/modulemd-module-stream.c [hunk 1]
```diff
@@ -684,3 +688,53 @@ static void
...
+                                        ModulemdTranslation *translation)
+{
+  g_return_if_fail (MODULEMD_IS_MODULE_STREAM (self));
+
+  ModulemdModuleStreamPrivate *priv =
+    modulemd_module_stream_get_instance_private (self);
+
+  g_clear_pointer (&priv->translation, g_object_unref);
+  if (translation != NULL)
+    priv->translation = modulemd_translation_copy (translation);
```
Comment (sgallagher) [inline modulemd/v2/modulemd-module-stream.c]: I'm really not sure we want to be copying this data around. My original thought was that we'd make these into references. In some module-stream-expansion cases, we may have MANY instances of the same module/stream pair in the Index. (Multiple versions, each with multiple contexts can grow quickly). I'd much rather we maintained a single copy at the ModuleIndex level that we associate to the individual streams via `g_object_ref()`. The summaries and descriptions can be arbitrarily long, and I'd rather not be duplicating hundreds of kilobytes of memory unnecessarily.

Also, if we move to referencing them, I think we should change the function name from `modulemd_module_stream_set_translation()` to `modulemd_module_stream_associate_translation()`, which will make it more clear that it's taking a reference and not making a copy.

File: modulemd/v2/modulemd-module.c [hunk 2]
```diff
@@ -205,10 +211,20 @@ modulemd_module_get_defaults (ModulemdModule *self)
 void
 modulemd_module_add_stream (ModulemdModule *self, ModulemdModuleStream *stream)
 {
+  ModulemdTranslation *translation = NULL;
+  ModulemdModuleStream *newstream = NULL;
   g_return_if_fail (MODULEMD_IS_MODULE (self));
 
-  g_ptr_array_add (self->streams,
-                   modulemd_module_stream_copy (stream, NULL, NULL));
+  //newstream = modulemd_module_stream_copy (stream, NULL, NULL);
```
Comment (sgallagher) [inline modulemd/v2/modulemd-module.c]: This should probably be removed.

File: modulemd/v2/modulemd-module.c [hunk 3]
```diff
@@ -205,10 +211,20 @@ modulemd_module_get_defaults (ModulemdModule *self)
...
+  newstream = g_object_ref (stream);
+  g_ptr_array_add (self->streams, newstream);
```
Comment (sgallagher) [inline modulemd/v2/modulemd-module.c]: Why did you move from a copy to a reference here?

Also, it just occurred to me that this function really ought to have
```
g_return_if_fail (g_str_equal(modulemd_module_stream_get_name(stream), self->module_name);
```
since it should really be a coding error if we ever try to add a stream whose name doesn't match the module.

File: modulemd/v2/modulemd-module.c [hunk 4]
```diff
@@ -293,3 +309,39 @@ modulemd_module_get_stream_by_NSVC (ModulemdModule *self,
...
+
+void
+modulemd_module_add_translation (ModulemdModule *self,
+                                 ModulemdTranslation *translation)
+{
+  gsize i;
+  ModulemdModuleStream *stream = NULL;
+
+  if (!g_str_equal (modulemd_translation_get_module_name (translation),
+                    modulemd_module_get_module_name (self)))
```
Comment (sgallagher) [inline modulemd/v2/modulemd-module.c]: This should be `g_return_if_fail()` so that it generates a log message and calls `abort()` when `G_DEBUG=fatal-warnings`. We should be able to assume that the caller should already have validated this and isn't just trying to assign translations at random. Or if they are, they should rightly get called out on it.

File: modulemd/v2/modulemd-module.c [hunk 5]
```diff
@@ -293,3 +309,39 @@ modulemd_module_get_stream_by_NSVC (ModulemdModule *self,
...
+    return;
+
+  g_hash_table_replace (
+    self->translations,
+    g_strdup (modulemd_translation_get_module_stream (translation)),
+    g_object_ref (translation));
```
Comment (sgallagher) [inline modulemd/v2/modulemd-module.c]: You're taking a reference to the translation here, but making a copy of it for each individual stream that is added. That really doesn't make any sense to me. We should make a copy onto the highest-level location that will need to reference it, then take a reference on it at each lower level. So in this case, I think we need to take the copy *here* and then have `modulemd_module_stream_associate_translation()` take a reference on each stream that matches.

Also, when implementing the ModuleIndex we need to have the addition of a Translation or Defaults object cause the creation of the appropriate Module object, if a Stream has not already done so.

File: modulemd/v2/modulemd-module.c (hunk 3 above)
Comment (puiterwijk) [reply modulemd/v2/modulemd-module.c]: The move from copy to ref is a result of the fact that the stream v1 and v2 objects don't yet have their own `copy` implementations, so that internal things (like the summary and description) don't get copied when you copy a stream.
Because of that I moved to a ref during development so I could at least verify that this specific patchset was working.
This patch probably should wait until we've added that, so that the tests can pass with a copy.

Comment (sgallagher) [reply modulemd/v2/modulemd-module.c]: Aargh, I didn't implement the `copy()`, did I? OK, I'll get that done by the end of Monday...

Comment (puiterwijk) [reply modulemd/v2/modulemd-module.c]: I actually have a fair bit of the `copy()` working for stream v1, so will finish that up and submit by Monday morning.

File: modulemd/v2/modulemd-module.c (hunk 5 above)
Comment (puiterwijk) [reply modulemd/v2/modulemd-module.c]: The creation of the Module object I had already done: `get_or_create_module`

**Comment by puiterwijk** [comment]
I updated this to fix the requested changes. It is now based on top of #143 to have a working `copy()`.

File: modulemd/v2/include/modulemd-2.0/private/modulemd-module-private.h [hunk 6]
```diff
@@ -53,6 +54,31 @@ modulemd_module_set_defaults (ModulemdModule *self,
                               ModulemdDefaults *defaults);
 
 
+/**
+ * modulemd_module_associate_translation:
+ * @self: This #ModulemdModule object
+ * @translation: A #ModulemdTranslation object to associate with this
+ * #ModulemdModule
```
Comment (sgallagher) [inline modulemd/v2/include/modulemd-2.0/private/modulemd-module-private.h]: GObject treats it as implicit here, but I'd prefer the `(in)` annotation for `@translation` as well as having the comment expressly state that it takes a reference and not a copy of the translation.

File: modulemd/v2/modulemd-module.c (hunk 4 above)
Comment (sgallagher) [reply modulemd/v2/modulemd-module.c]: You marked this "resolved" but didn't change it to `g_return_if_fail()` here. I'm guessing this was an oversight.

File: modulemd/v2/modulemd-module.c (hunk 5 above)
Comment (sgallagher) [reply modulemd/v2/modulemd-module.c]: We need to take a copy of the translation here (and use references to that copy within the associated streams). It's possible at the Index layer that we would get a Translation document passed in rather than built from YAML and we don't want to make the caller responsible for the memory (or for not changing it after it's referenced).

**Comment by puiterwijk** [comment]
Okay, I think I finally might have the memory ownership correct now...
